### DISCUSSION TITLE
Are there articles about moving from liburing to lower level system calls

### DISCUSSION THREAD

**Discussion by Ebenezer-group** [axboe/liburing#1179] [Q&A]
I'm thinking about trying to make this transition with [this program.](https://github.com/Ebenezer-group/onwards/blob/master/src/tiers/cmwA.cc)  The reason being to remove the dependency and improve build times.  I'm not sure how soon I'll be able to make this move so I also welcome thoughts on how to improve the current implementation.  Thanks.

**Comment by isilence** [comment]
> I'm thinking about trying to make this transition with [this program.](https://github.com/Ebenezer-group/onwards/blob/master/src/tiers/cmwA.cc) The reason being to remove the dependency and improve build times. I'm not sure how soon I'll be able to make this move so I also welcome thoughts on how to improve the current implementation. Thanks.

Oh, frankly I don't think it's a good idea at all. First it takes me 2.2 seconds to build a freshly cloned liburing without tests/examples with one thread. Parallelize it with `make -jN` and it's 1.1s, and it's that much only because configure is running slow (makes me wonder if that can be improved). Linking against it is also fast, while building any C++ always feels like ages comparing to that. 

Harder to be convincing with deps, but liburing is already a low[est] level library not doing much but forwarding to the kernel, while hiding a lot of boiler plate code inside. And it doesn't have any dependencies itself, not a dependency hell situation here.

With that being said, I have a feeling you won't find any articles, you'd just need to read through liburing and copy from `liburing.h` and `src*`, which would be lots of boiler plate and easy to screw up calculations.

**Reply by Ebenezer-group** [reply]
Thanks for your reply.  What about merging liburing into Linux somehow?

**Reply by isilence** [reply]
What do you mean? We can't move more of it into the kernel, liburing is already thin, there is no point in doing so, and even if we manage to move some parts there will still be enough of binding code left.

Just to entertain the idea, liburing can be vdso'd into the userspace, but IIRC that would be mapping liburing for every application even if they don't use io_uring, and having it separate is a much cleaner option anyway. And again there is not much point since it'll be doing just exactly what was done by a separate library, doing syscalls and such.
